### PR TITLE
fix: respect WillPopScope

### LINKED ISSUE #44: Add a way to prevent popping the first page of the FlowBuilder
**Is your feature request related to a problem? Please describe.**

Currently, the only way to show a back button on the first page of the `FlowBuilder` is to manually add it to the app bar as described [here](https://github.com/felangel/flow_builder/issues/36#issuecomment-784277855). In that case, `WillPopScope` included on the same page will not trigger after using either the app bar or Android back button. Similarly, neither `didPop` nor `didRemove` will be triggered when using a navigator observer passed to the `FlowBuilder`. This means that there is no easy way to observe and prevent popping the first page of the `FlowBuilder`. 

An example could be an onboarding page (as the first page of the `FlowBuilder`) where the Android back button should swipe back to the previous onboarding screen (but not pop the page). The problem does not occur if the onboarding page is the second page of the `FlowBuilder` with an app bar that uses a default back button - in that case `WillPopScope` can prevent the navigation back.

**Describe the solution you'd like**
It would be nice to have a similar structure as `WillPopScope`. I am not sure though about this solution, it could result in an incorrect navigation stack when `onPopPage` returns `false` after the flow being updated.

```dart
FlowBuilder(
  state: const FlowState(),
  onGeneratePages: (state, pages) {...},
  onComplete: (state) {...},
  onPopPage: (state, page) async {
    return shouldPop;
  }
)
```

**Describe alternatives you've considered**
- `WillPopScope`

```dart
FlowBuilder(
  state: const FlowState(),
  onGeneratePages: (state, pages) => [
    PageA.page(),
    ...
  ],
);
```

```dart
class PageA extends StatelessWidget {
  static Page page() => MaterialPage<void>(child: PageA());

  @override
  Widget build(BuildContext context) {
    return WillPopScope(
      onWillPop: () async {
        /// This will not trigger when either the app bar or Android back button is pressed.
        return false;
      },
      child: Scaffold(
        backgroundColor: Colors.yellow,
        appBar: AppBar(
          leading: BackButton(
            onPressed: () => context.flow<FlowState>().complete(),
          ),
          title: const Text('Page A'),
        ),
      ),
    );
  }
}
```

**Additional context**
None

### PR DESCRIPTION
**Pull request by felangel**
## Status

**READY**

## Breaking Changes

NO

## Description

- If there is a `WillPopScope` at the root of a flow, `FlowBuilder` will invoke and respect the result of `willPop`.
  - closes #44 

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [X] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore
- [ ] 🧪 Testing

### CODE REVIEW

**Comment by felangel** [comment]
@bselwe, @orestesgaolin, @jeroen-meijer can you take a look when you have time? Thanks! 🙏

**Review by orestesgaolin — APPROVED** [review]
This works as expected, thanks for this fix!

File: test/flow_builder_test.dart [hunk 1]
```diff
@@ -1182,6 +1182,118 @@ void main() {
...
+      await tester.tap(find.text('X'));
+      await tester.pumpAndSettle();
+
+      expect(find.byKey(targetKey), findsOneWidget);
+      await tester.tap(find.byKey(targetKey));
+      await tester.pumpAndSettle();
+
+      expect(onWillPopCallCount, equals(1));
+      expect(find.byKey(targetKey), findsNothing);
+    });
```
Comment (jeroen-meijer) [inline test/flow_builder_test.dart:1296]:
```suggestion
    },
    );
```

and formatting would look slightly prettier 😄

**Review by jeroen-meijer — APPROVED** [review]
Minor styling nit, but LGTM!
Nice fix :shipit:

**Review by bselwe — APPROVED** [review]
Looks good and works great with this fix, thanks!

File: test/flow_builder_test.dart (hunk 1 above)
Comment (felangel) [reply test/flow_builder_test.dart:1296]: If it’s okay with you I prefer to leave it in order to keep the test formatting consistent throughout.

**Pull request merged**